Metrictank crashes when asking for an aggregation that doesn't exist

For an aggregation configured to only compute `min` rollups, if a user asks for `avg` metrictank crashes:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x99bab4]
goroutine 344721736 [running]:
github.com/grafana/metrictank/api.doRecover(0xc5c88d1e48)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:29 +0x256
panic(0xbdd9a0, 0x115b350)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/panic.go:491 +0x283
github.com/grafana/metrictank/mdata.(*AggMetric).Get(0x0, 0x5a6cbee65a4531e6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/mdata/aggmetric.go:177 +0xb4
github.com/grafana/metrictank/mdata.(*AggMetric).GetAggregated(0xc48b9146e0, 0x2, 0x5a4531e600000e10, 0x5a6cbee6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/mdata/aggmetric.go:158 +0xea
github.com/grafana/metrictank/api.(*Server).getSeriesAggMetrics(0xc420782000, 0xc4e86c0550, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:483 +0x1c1
github.com/grafana/metrictank/api.(*Server).getSeries(0xc420782000, 0xc4e86c0550, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x10, ...)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:416 +0x7b
github.com/grafana/metrictank/api.(*Server).getSeriesFixed(0xc420782000, 0x1129400, 0xc8671ef980, 0xc5c4065c50, 0x22, 0xc7250dae60, 0x91, 0xc4ad331560, 0x85, 0x5a6cbee65a4531e6, ...)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:401 +0x132
github.com/grafana/metrictank/api.(*Server).getTarget(0xc420782000, 0x1129400, 0xc8671ef980, 0xc5c4065c50, 0x22, 0xc7250dae60, 0x91, 0xc4ad331560, 0x85, 0x5a6cbee65a4531e6, ...)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:340 +0x717
github.com/grafana/metrictank/api.(*Server).getTargetsLocal.func1(0x1129340, 0xc5e567aa00, 0xc420782000, 0xc5a7d39520, 0xc4ab8d8600, 0xc6bfe65880, 0xc448a6e660, 0xc5c4065c50, 0x22, 0xc7250dae60, ...)
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:274 +0x162
created by github.com/grafana/metrictank/api.(*Server).getTargetsLocal
	/Users/shanson7/Code/metrictank_notagsintree/src/github.com/grafana/metrictank/api/dataprocessor.go:270 +0x2bd
```

--- a/api/dataprocessor.go
+++ b/api/dataprocessor.go
@@ -413,7 +413,10 @@ func (s *Server) getSeriesFixed(ctx context.Context, req models.Req, consolidato
 }
 
 func (s *Server) getSeries(ctx *requestContext) (mdata.Result, error) {
-	res := s.getSeriesAggMetrics(ctx)
+	res, err := s.getSeriesAggMetrics(ctx)
+	if err != nil {
+		return res, err
+	}
 	select {
 	case <-ctx.ctx.Done():
 		//request canceled
@@ -468,14 +471,14 @@ func (s *Server) itersToPoints(ctx *requestContext, iters []chunk.Iter) []schema
 	return points
 }
 
-func (s *Server) getSeriesAggMetrics(ctx *requestContext) mdata.Result {
+func (s *Server) getSeriesAggMetrics(ctx *requestContext) (mdata.Result, error) {
 	_, span := tracing.NewSpan(ctx.ctx, s.Tracer, "getSeriesAggMetrics")
 	defer span.Finish()
 	metric, ok := s.MemoryStore.Get(ctx.Key)
 	if !ok {
 		return mdata.Result{
 			Oldest: ctx.Req.To,
-		}
+		}, nil
 	}
 
 	if ctx.Cons != consolidation.None {
@@ -483,7 +486,7 @@ func (s *Server) getSeriesAggMetrics(ctx *requestContext) mdata.Result {
 		return metric.GetAggregated(ctx.Cons, ctx.Req.ArchInterval, ctx.From, ctx.To)
 	} else {
 		logLoad("memory", ctx.Req.Key, ctx.From, ctx.To)
-		return metric.Get(ctx.From, ctx.To)
+		return metric.Get(ctx.From, ctx.To), nil
 	}
 }
 

--- a/api/dataprocessor_test.go
+++ b/api/dataprocessor_test.go
@@ -773,7 +773,10 @@ func TestGetSeriesAggMetrics(t *testing.T) {
 		metric.Add(i, float64(i^2))
 	}
 
-	res := srv.getSeriesAggMetrics(ctx)
+	res, err := srv.getSeriesAggMetrics(ctx)
+	if err != nil {
+		t.Errorf("Got unexpected error %q", err)
+	}
 	timestamps := make([]uint32, 0)
 	values := make([]float64, 0)
 

--- a/mdata/aggmetric.go
+++ b/mdata/aggmetric.go
@@ -144,9 +144,9 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
 			var agg *AggMetric
 			switch consolidator {
 			case consolidation.None:
-				panic("cannot get an archive for no consolidation")
+				return Result{}, fmt.Errorf("cannot get an archive for no consolidation")
 			case consolidation.Avg:
-				panic("avg consolidator has no matching Archive(). you need sum and cnt")
+				return Result{}, fmt.Errorf("avg consolidator has no matching Archive(). you need sum and cnt")
 			case consolidation.Cnt:
 				agg = a.cntMetric
 			case consolidation.Lst:
@@ -158,7 +158,7 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
 			case consolidation.Sum:
 				agg = a.sumMetric
 			default:
-				panic(fmt.Sprintf("AggMetric.GetAggregated(): unknown consolidator %q", consolidator))
+				return Result{}, fmt.Errorf("AggMetric.GetAggregated(): unknown consolidator %q", consolidator)
 			}
 			if agg == nil {
 				return Result{}, fmt.Errorf("Consolidator %q not configured", consolidator)

--- a/mdata/aggmetric.go
+++ b/mdata/aggmetric.go
@@ -137,27 +137,33 @@ func (a *AggMetric) getChunk(pos int) *chunk.Chunk {
 	return a.Chunks[pos]
 }
 
-func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) Result {
+func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) (Result, error) {
 	// no lock needed cause aggregators don't change at runtime
 	for _, a := range a.aggregators {
 		if a.span == aggSpan {
+			var agg *AggMetric
 			switch consolidator {
 			case consolidation.None:
 				panic("cannot get an archive for no consolidation")
 			case consolidation.Avg:
 				panic("avg consolidator has no matching Archive(). you need sum and cnt")
 			case consolidation.Cnt:
-				return a.cntMetric.Get(from, to)
+				agg = a.cntMetric
 			case consolidation.Lst:
-				return a.lstMetric.Get(from, to)
+				agg = a.lstMetric
 			case consolidation.Min:
-				return a.minMetric.Get(from, to)
+				agg = a.minMetric
 			case consolidation.Max:
-				return a.maxMetric.Get(from, to)
+				agg = a.maxMetric
 			case consolidation.Sum:
-				return a.sumMetric.Get(from, to)
+				agg = a.sumMetric
+			default:
+				panic(fmt.Sprintf("AggMetric.GetAggregated(): unknown consolidator %q", consolidator))
 			}
-			panic(fmt.Sprintf("AggMetric.GetAggregated(): unknown consolidator %q", consolidator))
+			if agg == nil {
+				return Result{}, fmt.Errorf("Consolidator %q not configured", consolidator)
+			}
+			return agg.Get(from, to), nil
 		}
 	}
 	panic(fmt.Sprintf("GetAggregated called with unknown aggSpan %d", aggSpan))

--- a/mdata/ifaces.go
+++ b/mdata/ifaces.go
@@ -12,5 +12,5 @@ type Metrics interface {
 type Metric interface {
 	Add(ts uint32, val float64)
 	Get(from, to uint32) Result
-	GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) Result
+	GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) (Result, error)
 }

